sled-agent/src/instance.rs

```diff
@@ -1521,3 +1525,656 @@ impl InstanceRunner {
+                storage_handle,
+                &temp_dir,
+            ),
+        )
+        .await
+        .expect("timed out creating Instance struct");
+
+        let (put_tx, put_rx) = oneshot::channel();
+
+        inst.put_state(put_tx, InstanceStateRequested::Running)
+            .await
+            .expect("failed to send Instance::put_state");
+
+        timeout(TIMEOUT_DURATION, put_rx)
+            .await
+            .expect("timed out waiting for Instance::put_state result")
+            .expect("failed to receive Instance::put_state result")
+            .expect("Instance::put_state failed");
+
+        timeout(
+            TIMEOUT_DURATION,
+            state_rx.wait_for(|maybe_state| match maybe_state {
+                ReceivedInstanceState::InstancePut(sled_inst_state) => {
+                    sled_inst_state.vmm_state.state == InstanceState::Running
+                }
+                _ => false,
+            }),
+        )
+        .await
+        .expect("timed out waiting for InstanceState::Running in FakeNexus")
+        .expect("failed to receive FakeNexus' InstanceState");
+
+        logctx.cleanup_successful();
+    }
+
+    // tests around dropshot request timeouts during the blocking propolis setup
+    #[tokio::test]
+    async fn test_instance_create_timeout_while_starting_propolis() {
+        let logctx = omicron_test_utils::dev::test_setup_log(
+            "test_instance_create_timeout_while_starting_propolis",
+        );
+
+        // automock'd things used during this test
+        let _mock_vnic_contexts = mock_vnic_contexts();
+        let _mock_zone_contexts = mock_zone_contexts();
+
+        let FakeNexusParts { nexus_client, nexus_server, state_rx } =
+            FakeNexusParts::new(&logctx);
+
+        let (_dns_server, resolver, _dns_config_dir) =
+            timeout(TIMEOUT_DURATION, dns_server(&logctx, &nexus_server))
+                .await
+                .expect("timed out making DNS server and Resolver");
+
+        let nexus_client_with_resolver =
+            NexusClientWithResolver::new_with_client(nexus_client, resolver);
+
+        let storage_handle = fake_storage_manager_with_u2().await;
+
+        let temp_guard = Utf8TempDir::new().unwrap();
+        let temp_dir = temp_guard.path().to_string();
+
+        let inst = timeout(
+            TIMEOUT_DURATION,
+            instance_struct(
+                &logctx,
+                // we want to test propolis not ever coming up
+                SocketAddr::V6(SocketAddrV6::new(Ipv6Addr::LOCALHOST, 1, 0, 0)),
+                nexus_client_with_resolver,
+                storage_handle,
+                &temp_dir,
+            ),
+        )
+        .await
+        .expect("timed out creating Instance struct");
+
+        let (put_tx, put_rx) = oneshot::channel();
+
+        inst.put_state(put_tx, InstanceStateRequested::Running)
+            .await
+            .expect("failed to send Instance::put_state");
+
```

Comment:
Can we call [`tokio::time::advance`](https://docs.rs/tokio/latest/tokio/time/fn.advance.html) here, so that we don't have to wait for the full `TIMEOUT_DURATION`? This will also allow us to increase that duration so without making the success case necessarily take longer, but still giving plenty of time so we don't get flaky runs in CI.

I think to do this, you'll have to create the timeout future, then advance the time, and then await the future.